File: packages/ifc-spatial-py/src/ifc_spatial/cache.py
"""Content-addressed model cache.

A port of ``packages/ifc-spatial/src/cache.ts`` and the ``SpatialCache`` at the
foot of ``model.ts``, merged into one class because on this engine there is only
one thing to cache. The TS package has two — a kilobyte-scale graph and a
megabyte-scale geometry index — and bounds them separately. IfcOpenShell holds
the file, the tessellations, the OCCT tree and the contact map behind a single
:class:`~ifc_spatial.model.SpatialModel`, so there is one lifetime and one lid.

## Why the key is the content and not the path

The premise is "just send the file": a host hands over bytes and gets a model
back, with no upload table and no lifecycle to keep in step. That premise only
survives contact with a conversation if the same bytes are never read twice — a
150 MB IFC costs seconds and several times its own size in peak memory, and an
agent that asks four questions about one building must not pay that four times.

The key is therefore the sha256 of the source bytes, not a file name, a path or
an upload id. Two consequences that are the point rather than side effects: a
re-upload of an unchanged file is a hit, and an edited file is a guaranteed
miss. Nothing here can serve a model that belongs to different bytes, which is
the failure a name-keyed cache eventually produces and cannot be talked out of.

Hashing is not free — a few hundred milliseconds for a large file — but it is an
order of magnitude below the read it decides. Measured on the sample house:
2.3 MB hashes in 2 ms against a 245 ms open, before the geometry pass that costs
another 1.6–2.4 s.

## Why the bound is an entry count and not a byte budget

The TS ``GraphCache`` carries an ``estimateBytes`` built from node and edge
counts, and its own docstring says the number "must never be reported as a
measurement". That estimate was defensible there because a graph is plain
JavaScript objects whose shape the code knows. Here it is not defensible at all:

  - the dominant terms are **outside Python**. The triangulations, the OCCT
    UB-tree and the BRep shapes live in the C++ heap behind SWIG proxies, and
    ``sys.getsizeof`` on the proxy returns the size of the pointer. A model
    measured at 351 MB RSS would be "estimated" at a few hundred kilobytes;
  - the cost is not proportional to anything countable up front. The same
    entity count produces wildly different memory depending on how curved the
    solids are, because the tessellator's chord deviation decides the triangle
    count, not the file;
  - the retained set GROWS after admission. ``SpatialModel`` is lazy: a model
    admitted after a topological question weighs almost nothing and the same
    object weighs hundreds of megabytes once someone asks for a clear height.
    A byte lid computed at admission would be wrong by the time it mattered.

An estimate that cannot be trusted is worse than an honest coarse limit, because
someone eventually puts it in a dashboard. So the lid is a count of models, kept
small on purpose — a conversation touches one building, occasionally two being
compared — and a host that needs a real memory bound should watch RSS and call
:meth:`SpatialCache.clear`, which is the only number that is actually true.

## What this cache is not

Process-local and unsynchronised. It holds no lock across processes, spans no
replica and survives no restart. It also has no tenancy: **a hash is not a
permission.** A host serving several tenants from one process must decide
whether a caller may see a building BEFORE it reaches this cache, because a
cache keyed on content happily serves anyone who can produce the same bytes.
"""

from __future__ import annotations

import atexit
import hashlib
import os
import tempfile
import threading
from collections import OrderedDict
from pathlib import Path

from .model import SpatialModel

#: Every cache's spill table, so the interpreter can remove the files it wrote
#: even though the cache itself must not (see ``SpatialCache._temp``).
_TEMP_FILES: list[dict[str, str]] = []


@atexit.register
def _remove_spilled_files() -> None:
    for table in _TEMP_FILES:
        for path in table.values():
            try:
                os.unlink(path)
            except OSError:
                pass


#: Read size for hashing. One megabyte keeps a 150 MB file off the heap without
#: turning the hash into a syscall storm.
_CHUNK = 1 << 20


def hash_bytes(data: bytes) -> str:
    """sha256 of these bytes, hex."""
    return hashlib.sha256(data).hexdigest()


def hash_file(path: str | os.PathLike[str]) -> str:
    """sha256 of this file's bytes, hex — streamed, never fully resident."""
    digest = hashlib.sha256()
    with Path(path).open("rb") as handle:
        for chunk in iter(lambda: handle.read(_CHUNK), b""):
            digest.update(chunk)
    return digest.hexdigest()


class _Pending:
    """One read in flight, and everyone waiting on it.

    Deliberately not :class:`concurrent.futures.Future`: this cache must work
    the same way whether it is called from a thread pool, from an asyncio
    handler via ``to_thread``, or from a plain synchronous test, and an
    ``Event`` plus two slots is the only shape that does all three without
    dragging an executor into the library.
    """

    __slots__ = ("done", "model", "error")

    def __init__(self) -> None:
        self.done = threading.Event()
        self.model: SpatialModel | None = None
        self.error: BaseException | None = None


class SpatialCache:
    """Opened models, keyed by the sha256 of their bytes."""

    def __init__(self, max_entries: int = 4) -> None:
        #: Insertion order IS the LRU order: a hit moves the key to the end, so
        #: the oldest key is always the first one iteration yields. No second
        #: structure, no timestamps to keep in step with the entries they
        #: describe.
        self._models: OrderedDict[str, SpatialModel] = OrderedDict()
        self._in_flight: dict[str, _Pending] = {}
        self._lock = threading.Lock()
        self._max_entries = max(1, max_entries)
        self._hits = 0
        self._misses = 0
        #: Temporary files this cache created for byte-sourced models.
        #:
        #: Never deleted on eviction, and that is deliberate. ``SpatialModel``
        #: holds the PATH, and IfcOpenShell reads the file again for the
        #: geometry pass and the OCCT tree — so an evicted model handed out
        #: earlier is still live in a caller's hands and still needs its file.
        #: Unlinking it would turn an eviction (a memory decision) into a broken
        #: model (a correctness one). They are cleaned up when the process ends.
        self._temp: dict[str, str] = {}
        _TEMP_FILES.append(self._temp)

    # ── introspection ───────────────────────────────────────────────────────

    @property
    def stats(self) -> dict[str, int]:
        """Counters for the process, not for a request.

        ``misses`` counts READS STARTED and ``hits`` counts calls served without
        one — including a call that joined a read already in flight. Defined
        that way because the number anyone actually wants from a cache is "how
        many times did we open an IFC", and a coalesced caller did not.
        """
        return {"hits": self._hits, "misses": self._misses, "entries": len(self._models)}

    def has(self, content_hash: str) -> bool:
        """Is this hash resident?

        Neither counted nor treated as a use: a question ABOUT the cache must
        not change what the cache would evict next, or a monitoring loop quietly
        becomes the thing that decides retention.
        """
        with self._lock:
            return content_hash in self._models

    def get(self, content_hash: str) -> SpatialModel | None:
        """The model for this hash, or ``None`` — the lookup a caller makes when
        it holds a handle and not the bytes."""
        with self._lock:
            model = self._models.get(content_hash)
            if model is None:
                self._misses += 1
                return None
            self._models.move_to_end(content_hash)
            self._hits += 1
            return model

    def resolve(self, prefix: str) -> SpatialModel | None:
        """A model by an abbreviated hash, the way a git object is addressed.

        Returns ``None`` for both "no such prefix" and "ambiguous prefix"; the
        caller distinguishes them with :meth:`hashes`, because the two need
        different German sentences and this class has no opinion about German.
        """
        key = prefix.strip().lower()
        with self._lock:
            if key in self._models:
                self._models.move_to_end(key)
                return self._models[key]
            matches = [h for h in self._models if h.startswith(key)] if len(key) >= 8 else []
        if len(matches) == 1:
            return self.get(matches[0])
        return None

    def hashes(self) -> list[str]:
        with self._lock:
            return list(self._models)

    # ── loading ─────────────────────────────────────────────────────────────

    def load(self, path: str | os.PathLike[str]) -> SpatialModel:
        """The model for this file — from the cache, from a read already
        running, or from a read this call starts. Exactly one of the three."""
        return self._load(hash_file(path), str(path), None)

    def load_bytes(self, data: bytes, *, suffix: str = ".ifc") -> SpatialModel:
        """The model for these bytes.

        The bytes are spilled to a temporary file, because ``ifcopenshell.open``
        wants a path and because the geometry iterator and the OCCT tree read
        the file again after the entity graph is in memory — a string handed to
        ``ifcopenshell.file.from_string`` would have to stay resident twice
        over, once as Python bytes and once inside the C++ reader. The spill
        happens once per content hash, so a re-upload of the same file costs
        nothing.
        """
        content = hash_bytes(data)
        with self._lock:
            existing = self._models.get(content)
            if existing is not None:
                self._models.move_to_end(content)
                self._hits += 1
                return existing
            path = self._temp.get(content)
        if path is None:
            handle, path = tempfile.mkstemp(prefix=f"ifc-spatial-{content[:12]}-", suffix=suffix)
            with os.fdopen(handle, "wb") as file:
                file.write(data)
            with self._lock:
                self._temp[content] = path
        return self._load(content, path, data)

    def _load(self, content: str, path: str, data: bytes | None) -> SpatialModel:
        with self._lock:
            resident = self._models.get(content)
            if resident is not None:
                self._models.move_to_end(content)
                self._hits += 1
                return resident
            running = self._in_flight.get(content)
            if running is not None:
                # Single-flight, and it matters more here than it does for a
                # graph: two concurrent opens of one model would run two
                # geometry passes and hold two full tessellations at the same
                # instant, which is precisely when memory is tightest. That is
                # not a rare race — it is the ordinary shape of a user
                # double-clicking, a retry landing on the same replica, or an
                # agent firing two tool calls at one building.
                self._hits += 1
                pending = running
            else:
                self._misses += 1
                pending = _Pending()
                self._in_flight[content] = pending
                running = None

        if running is not None:
            pending.done.wait()
            if pending.error is not None:
                raise pending.error
            assert pending.model is not None
            return pending.model

        try:
            model = SpatialModel(path)
        except BaseException as error:  # noqa: BLE001 — re-raised to every waiter
            pending.error = error
            with self._lock:
                # Cleared on BOTH outcomes. Leaving a failed read in the map
                # would poison the hash for the life of the process: every later
                # caller would join a read that already failed and receive its
                # error, turning one transient failure — a truncated upload, a
                # momentary allocation failure — into a permanent one for those
                # exact bytes.
                self._in_flight.pop(content, None)
            pending.done.set()
            raise
        else:
            pending.model = model
            with self._lock:
                # Admission on the success path only: a read that threw must
                # never leave a half-built model behind a cache hit.
                self._models[content] = model
                self._models.move_to_end(content)
                self._evict(keep=content)
                self._in_flight.pop(content, None)
            pending.done.set()
            return model

    def _evict(self, keep: str) -> None:
        """Evict oldest-first until the lid holds — but never the entry just
        admitted.

        A cache that can evict what it has only just stored re-reads the same
        file on every single call, which is worse than no cache: same read cost,
        plus the hashing, plus a stats line claiming a cache exists.
        """
        while len(self._models) > self._max_entries:
            oldest = next(iter(self._models))
            if oldest == keep:
                break
            self._models.pop(oldest, None)

    def clear(self) -> None:
        """Drop every retained model.

        Reads already in flight are left running and their results are still
        admitted. They are not ours to cancel — callers are waiting on them —
        and a model is valid for its hash no matter when it was asked for, so
        refusing to admit it would only guarantee the next caller re-reads the
        same file.

        ``hits`` and ``misses`` survive: they describe the process's history,
        and resetting them on a clear would hide the very re-reads a clear
        causes.
        """
        with self._lock:
            self._models.clear()


__all__ = ["SpatialCache", "hash_bytes", "hash_file"]

File: packages/ifc-spatial/src/cache.ts
/**
 * Content-addressed graph cache.
 *
 * ## Why this exists at all
 *
 * The whole premise of this library is "just send the file": a host hands over
 * bytes and gets a {@link BuildingGraph} back, with no model id, no upload
 * table, no lifecycle to keep in step. That premise only survives contact with
 * a conversation if the same bytes are never parsed twice. A 150 MB IFC costs
 * seconds and several times its own size in peak memory to read; an agent that
 * asks four questions about one building must not pay that four times.
 *
 * The cache key is therefore the sha256 of the source bytes — the same hash the
 * builder already writes to {@link BuildingGraph.contentHash} — and not a file
 * name, a path or an upload id. Two consequences that are the point rather than
 * side effects: a re-upload of an unchanged file is a hit, and an edited file is
 * a guaranteed miss. Nothing here can serve a graph that belongs to different
 * bytes, which is the failure a name-keyed cache eventually produces and cannot
 * be talked out of.
 *
 * Hashing is not free — a few hundred milliseconds for a large file — but it is
 * an order of magnitude below the parse it decides, and it is the only way to
 * be certain the graph in hand describes the file in hand.
 *
 * ## What this cache is not
 *
 * Process-local and unsynchronised. It holds no lock, spans no replica and
 * survives no restart. A host that wants a shared or persistent cache should
 * put one in front of {@link GraphCache.load} rather than inside it: this class
 * exists to stop one process from re-parsing what it already holds, and the
 * moment it tries to be a distributed cache it acquires a coherence problem it
 * has no way to solve.
 *
 * It also has no tenancy. A hash is not a permission — a host that serves
 * several tenants from one process must decide whether a caller may see a
 * building BEFORE it reaches this cache, because a cache keyed on content
 * happily serves anyone who can produce the same bytes.
 */

import { createHash } from 'node:crypto'
import { buildGraph, type BuildOptions } from './graph/build.js'
import type { BuildingGraph } from './graph/types.js'

export interface GraphCacheOptions {
  /**
   * Hard lid on retained graphs. Small on purpose: a conversation touches one
   * building, occasionally two being compared, and the value of a third resident
   * graph is far below the cost of holding it.
   */
  maxEntries?: number
  /**
   * Lid on the ESTIMATED retained size (see {@link estimateBytes}). Present
   * because entry count alone cannot tell a sample house from a federated
   * hospital, and eight of the latter is not a cache, it is a leak.
   */
  maxBytes?: number
}

interface CacheEntry {
  graph: BuildingGraph
  /** Estimated, never measured. See {@link estimateBytes}. */
  bytes: number
}

/**
 * Rough per-object costs for the size estimate.
 *
 * These are shape arithmetic, not measurements: a `GraphNode` is eleven fields
 * of which most are heap strings the parser handed us (a GlobalId alone is 22
 * characters plus a string header), and an edge is charged more than its own
 * object because every edge also occupies two adjacency slots and contributes
 * Map keys. The numbers are deliberately round and deliberately generous — an
 * estimate that runs low is one that lets the cache overshoot the memory the
 * host actually has.
 */
const BYTES_PER_NODE = 320
const BYTES_PER_EDGE = 200
const BYTES_PER_DIALECT_ENTRY = 400
const BYTES_BASE = 64 * 1024

/**
 * Approximate retained size of a graph, in bytes.
 *
 * **This is an estimate and must never be reported as a measurement.** It is
 * derived from {@link BuildingGraph.stats} counts, so it knows nothing about
 * string interning, V8's actual object layout, or the fact that node names in a
 * German model are longer than in an English one. Its only job is to order
 * evictions and to stop the cache growing without bound; anything else — a
 * memory figure in a UI, a capacity plan, a billing number — is a misuse, and
 * the moment someone treats it as truth the estimate becomes a lie with a
 * decimal point on it.
 */
export function estimateBytes(graph: BuildingGraph): number {
  return (
    BYTES_BASE +
    graph.stats.nodes * BYTES_PER_NODE +
    graph.stats.edges * BYTES_PER_EDGE +
    graph.dialect.length * BYTES_PER_DIALECT_ENTRY
  )
}

/** sha256 of the bytes, hex — identical to what {@link buildGraph} stamps on the graph. */
export function hashBytes(bytes: Uint8Array): string {
  return createHash('sha256').update(bytes).digest('hex')
}

export class GraphCache {
  /**
   * Insertion order IS the LRU order: a hit deletes and re-inserts, so the
   * oldest key is always the first one `Map` iteration yields. No second
   * structure, no timestamps to keep in step with the entries they describe.
   */
  readonly #entries = new Map<string, CacheEntry>()

  /**
   * Parses currently running, keyed by hash — the single-flight table.
   *
   * Two concurrent `load()` calls for the same bytes must produce ONE parse.
   * Without this map they produce two, and the two peak at the same instant:
   * a 150 MB model parsed twice concurrently costs twice the peak memory at
   * exactly the moment memory is tightest, which is how a process that handles
   * one big model comfortably dies on two identical requests that arrive a
   * hundred milliseconds apart. That is not a rare race — it is the ordinary
   * shape of a user double-clicking, a retry landing on the same replica, or an
   * agent firing two tool calls at one building.
   *
   * Note this deduplicates by CONTENT, so it also collapses the case the caller
   * cannot see: two different requests that happen to carry the same file.
   */
  readonly #inFlight = new Map<string, Promise<BuildingGraph>>()

  readonly #maxEntries: number
  readonly #maxBytes: number

  #hits = 0
  #misses = 0
  #bytes = 0

  constructor(options: GraphCacheOptions = {}) {
    this.#maxEntries = Math.max(1, options.maxEntries ?? 8)
    this.#maxBytes = Math.max(1, options.maxBytes ?? 512 * 1024 * 1024)
  }

  /**
   * Counters for the process, not for a request.
   *
   * `misses` counts PARSES STARTED, and `hits` counts calls served without one —
   * including a call that joined a parse already in flight. Defined that way
   * because the number anyone actually wants from a cache is "how many times did
   * we read an IFC", and a coalesced caller did not.
   *
   * `bytes` is the estimate described in {@link estimateBytes}. It is a
   * snapshot: reading it twice around a `load()` is the only honest way to see
   * what that load cost, and even then only approximately.
   */
  get stats(): { hits: number; misses: number; entries: number; bytes: number } {
    return { hits: this.#hits, misses: this.#misses, entries: this.#entries.size, bytes: this.#bytes }
  }

  /**
   * The graph for `hash`, or `null`.
   *
   * This is the lookup a caller makes when it already has a hash and does NOT
   * have the bytes — "do I need to ask the user to re-upload". It records a hit
   * or a miss, because it is a real cache lookup; `load()` deliberately does not
   * go through it, so a load can never be counted twice.
   */
  get(hash: string): BuildingGraph | null {
    const entry = this.#touch(hash)
    if (entry) {
      this.#hits++
      return entry.graph
    }
    this.#misses++
    return null
  }

  /**
   * Is `hash` resident?
   *
   * Neither counted nor treated as a use: a question ABOUT the cache must not
   * change what the cache would evict next, or a monitoring loop quietly becomes
   * the thing that decides retention.
   */
  has(hash: string): boolean {
    return this.#entries.has(hash)
  }

  /**
   * The graph for these bytes — from the cache, from a parse already running, or
   * from a parse this call starts. Exactly one of the three.
   *
   * `options` are honoured only by the call that actually parses. A hit returns
   * the graph as it was built, which means a caller that passes a different
   * `dialectSampleSize` for the same bytes gets the earlier sampling. That is
   * the correct trade — re-parsing 150 MB to re-measure a dialect sample is not
   * a bargain anyone would take knowingly — but it is worth knowing, and a host
   * that needs different build options for the same file should use a second
   * cache instance rather than hoping.
   */
  async load(bytes: Uint8Array, options: BuildOptions = {}): Promise<BuildingGraph> {
    const hash = hashBytes(bytes)

    const resident = this.#touch(hash)
    if (resident) {
      this.#hits++
      return resident.graph
    }

    const running = this.#inFlight.get(hash)
    if (running) {
      this.#hits++
      return running
    }

    this.#misses++

    // The stored promise is the one every caller receives, so a rejection is
    // observed by all of them and by nothing else. Admission happens on the
    // success path only: a parse that threw must never leave a half-built or
    // absent graph behind a cache hit, because the next caller would then be
    // served a failure it cannot see and cannot retry out of.
    const parse = buildGraph(bytes, options).then((graph) => {
      this.#admit(hash, graph)
      return graph
    })
    // The in-flight entry is cleared on BOTH outcomes. Leaving a rejected
    // promise in the map would poison the hash for the life of the process:
    // every later caller would join a parse that already failed and receive its
    // error, turning one transient failure — a truncated upload, a momentary
    // allocation failure — into a permanent one for those exact bytes.
    const tracked = parse.finally(() => {
      this.#inFlight.delete(hash)
    })

    this.#inFlight.set(hash, tracked)
    return tracked
  }

  /**
   * Drop every retained graph.
   *
   * Parses already in flight are left running and their results are still
   * admitted. They are not ours to cancel — callers are waiting on them — and a
   * graph is valid for its hash no matter when it was asked for, so refusing to
   * admit it would only guarantee the next caller re-parses the same file. What
   * `clear()` releases is what the cache is holding, which is the only memory it
   * controls.
   *
   * `hits` and `misses` survive: they describe the process's history, and
   * resetting them on a clear would hide the very re-parses a clear causes.
   */
  clear(): void {
    this.#entries.clear()
    this.#bytes = 0
  }

  /** Look up and promote to most-recently-used in one step. */
  #touch(hash: string): CacheEntry | null {
    const entry = this.#entries.get(hash)
    if (!entry) return null
    this.#entries.delete(hash)
    this.#entries.set(hash, entry)
    return entry
  }

  #admit(hash: string, graph: BuildingGraph): void {
    const bytes = estimateBytes(graph)
    const existing = this.#entries.get(hash)
    if (existing) this.#bytes -= existing.bytes
    this.#entries.delete(hash)
    this.#entries.set(hash, { graph, bytes })
    this.#bytes += bytes
    this.#evict(hash)
  }

  /**
   * Evict oldest-first until both lids hold — but never the entry just admitted.
   *
   * A cache that can evict what it has only just stored re-parses the same file
   * on every single call, which is worse than no cache: same parse cost, plus
   * the hashing, plus a stats line claiming a cache exists. So a single graph
   * larger than `maxBytes` is kept, alone, and the limit is treated as a
   * high-water mark rather than a promise.
   */
  #evict(keep: string): void {
    for (const [hash, entry] of this.#entries) {
      if (this.#entries.size <= this.#maxEntries && this.#bytes <= this.#maxBytes) return
      if (hash === keep) continue
      this.#entries.delete(hash)
      this.#bytes -= entry.bytes
    }
  }
}
